server/routerlicious/packages/lambdas/src/test/deli/checkpointContext.spec.ts

```diff
@@ -52,9 +53,9 @@ describe("Routerlicious", () => {
 
             beforeEach(() => {
                 testContext = new testUtils.TestContext();
-                testCollection = new testUtils.TestCollection([{ documentId: testId, tenantId: testTenant }]);
-
-                const checkpointManager = createDeliCheckpointManagerFromCollection(testTenant, testId, testCollection);
+                testDocumentRepository = new testUtils.TestDocumentRepository();
+				Sinon.replace(testDocumentRepository, "updateDocument", Sinon.fake());
```

Comment:
Why do we have to stub this function? The stub doesn't appear to be used anywhere, and we're already using a Test* implementation of the class. Can we not just make the Test implementation work appropriately?

Edit: I see that the Test implementation only throws errors. I'm not familiar with this pattern, or at least haven't seen it in a while! Would you mind explaining the benefit of doing it this way, rather than implementing a TestDocumentRepository with a TestCollection for the documentCollection?

Reply:
Hey @znewton 

I came from a c# .net world so not quite sure if I am doing the right way. But for c# .net world (I think it should be same for any OOP):

1. We use inverse of control so we could inject dependencies,
2. We program to abstract so we can switch/change implementations

With these principles, it become easier for us to Mock the dependencies. in C# .net, I use [Moq library](https://github.com/moq/moq4) most. It can mock an interface or an abstract class, but not a concert class easily. So I think by creating a non implemented and throw class in node, we can benefits:

1. Avoid dependency hell. Consider some crazy scenarios that you have to create so many unrelated dependencies  to just mock one simple behavior
2. It also abstract away the mongo dependencies.
3. I on purposely throw for each method, so the unit test will fail for the places you use it. That will force people to pay attention to their program workflow and provide correct flow based dependencies' output.

That is why I decide to use a fake `TestDocumentRepository` and leave it independent from the database. I am not an expert of nodes world, is this pattern sounds a concern?

Reply:
Awesome! I like it. Thanks for explaining!